File: acwing/蓝桥杯每日一题2025/week2/5439.农夫约翰真的种地/5439.农夫约翰真的种地.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <cstdio>
#include <cmath>

using namespace std;

const int N = 2 * 1e5 + 10;
int h[N], a[N], rk[N];

int main()
{
    int T;
    scanf("%d", &T);
    while (T--)
    {
        int n;
        scanf("%d", &n);
        for (int i = 1; i <= n; i++)
            scanf("%d", &h[i]);
        for (int i = 1; i <= n; i++)
            scanf("%d", &a[i]);
        for (int i = 1; i <= n; i++)
        {
            int t;
            scanf("%d", &t);
            rk[t + 1] = i;
        }
        int l = 0, r = 1e9;
        for (int i = 1; i < n; i++)
        {
            int A = h[rk[i]] - h[rk[i + 1]];
            int B = a[rk[i + 1]] - a[rk[i]];

            if (B > 0)
            {
                r = min(r, (int)ceil((double)A / B) - 1); //注意这里不能用floor(A/B )当 A /B是整数是，这里只有ceil((double)A / B) - 1
            }
            else if (B < 0)
            {
                l = max(l, (int)floor((double)A / B) + 1);
            }
            else if (A <= 0)
            {
                r = -1;
                break;
            }
        }
        if (l > r)
            l = -1;
        printf("%d\n", l);
    }

    return 0;
}

File: acwing/蓝桥杯每日一题2025/week2/5439.农夫约翰真的种地/5439.农夫约翰真的种地.py
from math import ceil, floor

N = 2 * int(1e5) + 5
rk = [0] * N

T = int(input())

for _ in range(T):
    n = int(input())
    h = [0] + list(map(int, input().split()))
    a = [0] + list(map(int, input().split()))

    t = [0] + list(map(int, input().split()))
    for i in range(1, n + 1):
        rk[t[i] + 1] = i

    l = 0
    r = int(1e9)
    for i in range(1, n):
        A = h[rk[i]] - h[rk[i + 1]]
        B = a[rk[i + 1]] - a[rk[i]]

        if B > 0:
            r = min(r, ceil(A / B) - 1)
        elif B < 0:
            l = max(l, floor(A / B) + 1)
        elif A <= 0:
            r = -1
            break

    if l > r:
        l = -1
    print(l)
